src/shared/db.ts

```diff
@@ -1,3 +1,21 @@
 import { PrismaClient } from "@prisma/client";
+import { AdapterUser } from "next-auth/adapters";
 
 export const prisma = new PrismaClient();
+
+export async function createUser(loginUser: AdapterUser) {
+  const existingUser = await prisma.user.findUnique({
+    where: { email: loginUser.email },
+  });
+
```

Comment:
there should be some type of error handling
> idk how you guys plan to do so, but i think this func should try to create user, if fail it throws. and then who ever calls it, calls in a try catch.
e.g

if(existingUser) throw new Error('Email already in use')

Reply:
Do you think this is enough? Should I` throw new Error` on the `catch`? I'm not really sure on how to do this, console.log doesn't seem right. But I don't know what to Throw as an Error either hahah

`export async function createUser(loginUser: User) {
  try {
    const existingUser = await prisma.user.findUnique({
      where: { email: loginUser.email as string },
    });

    if (existingUser) {
      throw new Error("E-mail already in use");
    } else {
      const createdUser = await prisma.user.create({
        data: {
          email: loginUser.email as string,
          avatar: loginUser.image as string,
          name: loginUser.name as string,
          username: loginUser.email as string,
        },
      });

      if (!createdUser) {
        throw new Error("Server failed to create user");
      }
    }
  } catch (error) {
    console.error("Error creating user:", error);
  }
}`

Reply:
the function literally throws and you catch when you invoke it, or you could use promises and return the promise.
```ts
export async function createUser(loginUser: User) {
    const existingUser = await prisma.user.findUnique({
      where: { email: loginUser.email as string },
    });

    if (existingUser) {
      throw new Error("E-mail already in use");
    } else {
      const createdUser = await prisma.user.create({
        data: {
          email: loginUser.email as string,
          avatar: loginUser.image as string,
          name: loginUser.name as string,
          username: loginUser.email as string,
        },
      });

      if (!createdUser) {
        throw new Error("Server failed to create user");
      }
    }
}
```

Reply:
Not sure, if we need the last if check, as the await create would also throw an error if it fails

```ts
export async function createUser(loginUser: AdapterUser): User {
    const existingUser = await prisma.user.findUnique({
      where: { email: loginUser.email },
    });

    if (existingUser) {
      throw new Error("E-mail already in use");
    } else {
      return await prisma.user.create({
        data: {
          email: loginUser.email as string,
          avatar: loginUser.image as string,
          name: loginUser.name as string,
          username: loginUser.email as string,
        },
      });
    }
}
```

Reply:
I realized we can't throw an error if there's already an user in the DB as this would prevent users who already have an account in the DB from signing in. This is the state of this parte of the code, so far:

```typescript
export async function createUser(loginUser: User) {
  const existingUser = await prisma.user.findUnique({
    where: { email: loginUser.email as string },
  });

  if (!existingUser) {
    const createdUser = await prisma.user.create({
      data: {
        email: loginUser.email as string,
        avatar: (loginUser.image as string) ?? null,
        name: (loginUser.name as string) ?? (loginUser.email as string),
        username: loginUser.email as string,
      },
    });

    if (!createdUser) {
      throw new Error("Server failed to create user");
    }
  }
}
```

Reply:
I think we can validate this by creating some tests, but I think the DB throws errors when some unique key is duplicated, or when you try to create a user and it fails